lib/PAUSE/Indexer/Context.pm

```diff
@@ -0,0 +1,172 @@
+use PAUSE::Logger '$Logger';
+
+has package_warnings => (
+  is => 'bare',
+  reader  => '_package_warnings',
+  default => sub {  {}  },
+);
+
+sub add_package_warning {
+  my ($self, $package_obj, $warning) = @_;
+
+  my $package = $package_obj->{PACKAGE};
+  my $pmfile  = $package_obj->pmfile->{PMFILE};
+
+  my $key = "$package\0$pmfile";
+
+  my $list = ($self->_package_warnings->{$key} //= []);
+  push @$list, {
+    package => $package,
+    pmfile  => $pmfile,
+    text    => $warning,
+  };
+
+  $Logger->log([
+    "adding package warning to %s: %s",
+    $package_obj->{PACKAGE},
+    $list->[-1],
+  ]);
+
+  return;
+}
+
+has package_status => (
+  is => 'bare',
+  reader  => '_package_status',
+  default => sub {  {}  },
+);
+
+sub _set_package_error {
+  my ($self, $package_obj, $status) = @_;
+
+  $self->_package_status->{ $package_obj->{PACKAGE} } = {
+    is_success  => 0,
+    filename    => $package_obj->{PP}{infile},
+    version     => $package_obj->{PP}{version},
+    header      => $status->{header},
+    body        => $status->{body},
+    package     => $package_obj->{PACKAGE},
+  };
+
+  $Logger->log([
+    "set error status for %s",
+    $package_obj->{PACKAGE},
+  ]);
+
+  return;
+}
+
+sub record_package_indexing {
+  my ($self, $package_obj) = @_;
+
+  $self->_package_status->{ $package_obj->{PACKAGE} } = {
+    is_success  => 1,
+    filename    => $package_obj->{PP}{infile},
+    version     => $package_obj->{PP}{version},
+    header      => "Indexed successfully",
+    body        => "The package was indexed successfully.",
+    package     => $package_obj->{PACKAGE},
+  };
+
+  $Logger->log([
+    "set OK status for %s",
+    $package_obj->{PACKAGE},
+  ]);
+
+  return;
+}
+
+sub package_statuses {
+  my ($self) = @_;
+
+  my %stash = %{ $self->_package_status };
+  return @stash{ sort keys %stash };
+}
+
+sub abort_indexing_package {
+  my ($self, $package_obj, $error) = @_;
+
+  $Logger->log("abort indexing $package_obj->{PACKAGE}");
+
+  $self->_set_package_error($package_obj, $error);
+
+  die PAUSE::Indexer::Abort::Package->new({
+    message => $error->{header},
+    public  => 1,
+  });
+}
+
+sub warnings_for_all_packages {
+  my ($self) = @_;
+
+  return map {; @$_ } values %{ $self->_package_warnings };
+}
+
+sub warnings_for_package {
+  my ($self, $package_name) = @_;
+
+  return grep {; $_->{package} eq $package_name }
+         $self->warnings_for_all_packages;
+}
+
+has alerts => (
+  is      => 'bare',
+  reader  => '_alerts',
+  default => sub {  []  },
+);
+
+sub alert {
+  my ($self, $alert) = @_;
+  $alert =~ s/\v+\z//;
+
+  push @{ $self->_alerts }, $alert;
+  return;
+}
+
+sub all_alerts {
+  my ($self) = @_;
+  return @{ $self->_alerts };
+}
+
+has dist_errors => (
+  is      => 'bare',
+  reader  => '_dist_errors',
+  default => sub {  []  },
+);
+
+sub add_dist_error {
+  my ($self, $error) = @_;
+
+  $error = ref $error ? $error : { ident => $error, message => $error };
```

Comment:
Later code does:
```
  my @dist_errors = $ctx->dist_errors;

  for my $error (@dist_errors) {
    my $header = $error->{header};
    my $body   = $error->{body};
```

But if you say `add_dist_error("foobar")`, there will be no header or body. We should probably just drop the string support and always expect the object/hashref

Reply:
I think you're confused, but maybe I'm confused.

The ternary expression says "if $error was a reference, leave it one.  otherwise, turn it into one where it's both ident and message are the string that was $error".  So, whether you pass a string or reference to `add_dist_error`, you get a reference stored in `all_dist_errors`.  The "later code" you cite will get those references and treat them as references.

Reply:
This doesn't have to do with whether or not later code expects references, it's what that later code expects the references to **contain**.

As I said, later code wants a `header` and a `body` key, and doesn't care at all about `ident` or `message`. Here's the full snippet of what later code wants: 

```
  my @dist_errors = $ctx->all_dist_errors;

  for my $error (@dist_errors) {
    my $header = $error->{header};
    my $body   = $error->{body};
    $body = $body->($self) if ref $body;

    push @m, "## $header\n\n";
    push @m, $tf->format($body), qq{\n\n};

    $status_over_all = "Failed";
  }
```

If add_dist_error is called with a string, and we ever somehow do a mail summary send, it seems to me we'll push:
```
## \n\n

```
(for header and body respectively, as they'll be undef)
and generate two warnings.

Reply:
Relatedly, if we ever do send a public summary, it seems like we don't attempt to filter out non-public errors! We shoudl?

Reply:
Okay, I've fixed this.